### DISCUSSION TITLE
Both HTTP and HTTPS at the same time

### DISCUSSION THREAD

**Discussion by easymoney322** [gogs/gogs#7653] [Q&A]
Hello,
I'd like to know if its possible to have HTTPS and HTTP servers running at the same time without proxies. The reason is that I'm going to use a self-signed certificate. Some clients (on which I cannot install CA certificate) may require additional flags to be passed in order to ignore the certificate check otherwise, and its easier to fallback to http. 
 
Judging by custom/conf/app.ini configurations FAQ, it looks like these options are mutually exclusive. Is it posible to have multiple [server] sections? Or is there anyway to achieve this behaviour? Adding another webserver on the same machine just to proxypass requests to the localhost sounds like an overkill.

**Comment by annpocoyo** [comment]
Doesn't this completely defeat the purpose of HTTPS. If you are using a self signed certificate it might be easier to just use HTTP.
Also if the site is on the internet you can get free ssl certificates from [Let's Encrypt](https://letsencrypt.org).

**Reply by easymoney322** [reply]
No, it doesn't defeat its purpose. And most of the times its a case of a server that serves both internet and intranet.

1) If I run git client on the same physical host in a container or a vm, or with isolated\VLANed network\VPN gateway, it still takes extra-steps to install trust to the self-signed CA, while the connection itself is non-interceptable

2) If I terminate the connection on an air-gapped balancer for intranet, but I don't want to expose the balancer to the internet, while still allowing https clients from the internet (via DNAT).

Of course, I can just setup another nginx instance on the same host with gogs, run gogs on http with unprivileged port and restrict this port to allow connections from the localhost, then setup balancer for the use of url-based balancing and use two different certificates for each of URLs, and also only 301-redirecting URLs to HTTPS (but only the requests that aren't from the containers and VMs). 
All this + cache managment + extra security for balancer + extra server resources, just because devs don't want to allow the user to use both http+https at the same time.
Or just allow http+https to be used at the same time.

**Reply by annpocoyo** [reply]
Yeh but the problem with both http and https on the same server is that if the client for some reason chooses http first while there is a man in the middle then the connection is compromised and the client might just not use https, and if they choose https then the man in the middle could just block that and the client would just think https isn't available.

However, if you want you can create a feature request in the issues tab and maybe the developers would listen.
Also Gogs's builtin web server isn't very advanced so most people use a proxy.

**Reply by easymoney322** [reply]
True about trying http before https. However, IMO, its a job of a proxy or balancer to "upgrade" (e.g. proxy_set_header Upgrade) from http to https. I believe, HSTS and auto-upgrades are even disabled by default in IIS-web and NginX web-servers.

After all, we can just add feature flags or configuration options to allow both cases of http+https configurations, by making the auto-upgrade as the default option. Customization=good.

I'd already made a PR, but, unfortunately, I'm not familiar with GO-lang.

**Reply by annpocoyo** [reply]
If you access the Gogs server with a domain then you could get a free proper ssl certificate from Let's Encrypt, otherwise you will just have to use proxies and stuff for now.
